File: android-app/src/main/java/com/prevengos/plug/android/di/AppContainer.java
package com.prevengos.plug.android.di;

import android.content.Context;

import com.prevengos.plug.android.BuildConfig;
import com.prevengos.plug.android.data.local.PrevengosDatabase;
import com.prevengos.plug.android.data.remote.api.PrevengosSyncApi;
import com.prevengos.plug.android.data.repository.CuestionarioRepository;
import com.prevengos.plug.android.data.repository.PacienteRepository;
import com.prevengos.plug.android.data.repository.SyncRepository;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

public class AppContainer {
    private final PrevengosDatabase database;
    private final OkHttpClient okHttpClient;
    private final Retrofit retrofit;
    private final PrevengosSyncApi syncApi;
    private final ExecutorService ioExecutor;

    private final PacienteRepository pacienteRepository;
    private final CuestionarioRepository cuestionarioRepository;
    private final SyncRepository syncRepository;

    public AppContainer(Context context) {
        database = PrevengosDatabase.instance(context);
        ioExecutor = Executors.newSingleThreadExecutor();
        okHttpClient = buildHttpClient(BuildConfig.APPLICATION_ID);
        retrofit = new Retrofit.Builder()
                .baseUrl(resolveBaseUrl(BuildConfig.SYNC_BASE_URL))
                .addConverterFactory(MoshiConverterFactory.create())
                .client(okHttpClient)
                .build();
        syncApi = retrofit.create(PrevengosSyncApi.class);
        pacienteRepository = new PacienteRepository(database.pacienteDao());
        cuestionarioRepository = new CuestionarioRepository(database.cuestionarioDao());
        syncRepository = new SyncRepository(
                database.pacienteDao(),
                database.cuestionarioDao(),
                database.syncMetadataDao(),
                syncApi,
                BuildConfig.APPLICATION_ID);
    }

    private OkHttpClient buildHttpClient(String clientId) {
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BASIC);
        return new OkHttpClient.Builder()
                .addInterceptor(chain -> chain.proceed(
                        chain.request()
                                .newBuilder()
                                .addHeader("X-Source-System", clientId)
                                .build()))
                .addInterceptor(logging)
                .build();
    }

    public PacienteRepository getPacienteRepository() {
        return pacienteRepository;
    }

    public CuestionarioRepository getCuestionarioRepository() {
        return cuestionarioRepository;
    }

    public SyncRepository getSyncRepository() {
        return syncRepository;
    }

    public ExecutorService getIoExecutor() {
        return ioExecutor;
    }

    private String resolveBaseUrl(String candidate) {
        if (candidate == null || candidate.isEmpty()) {
            return "https://api.prevengos.test/";
        }
        return candidate.endsWith("/") ? candidate : candidate + "/";
    }
}

File: android-app/src/main/java/com/prevengos/plug/android/di/AppContainer.kt
package com.prevengos.plug.android.di

import android.content.Context
import com.prevengos.plug.android.data.local.PrevengosDatabase
import com.prevengos.plug.android.data.remote.api.PrevengosSyncApi
import com.prevengos.plug.android.data.repository.CuestionarioRepository
import com.prevengos.plug.android.data.repository.PacienteRepository
import com.prevengos.plug.android.data.repository.SyncRepository
import kotlinx.coroutines.Dispatchers
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

class AppContainer(context: Context) {
    private val database: PrevengosDatabase = PrevengosDatabase.instance(context)

    private val okHttpClient: OkHttpClient by lazy {
        val logging = HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BASIC
        }
        OkHttpClient.Builder()
            .addInterceptor(logging)
            .build()
    }

    private val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .client(okHttpClient)
            .build()
    }

    private val syncApi: PrevengosSyncApi by lazy {
        retrofit.create(PrevengosSyncApi::class.java)
    }

    val pacienteRepository: PacienteRepository by lazy {
        PacienteRepository(database.pacienteDao())
    }

    val cuestionarioRepository: CuestionarioRepository by lazy {
        CuestionarioRepository(database.cuestionarioDao())
    }

    val syncRepository: SyncRepository by lazy {
        SyncRepository(
            pacienteDao = database.pacienteDao(),
            cuestionarioDao = database.cuestionarioDao(),
            syncMetadataDao = database.syncMetadataDao(),
            syncApi = syncApi,
            ioDispatcher = Dispatchers.IO
        )
    }

    companion object {
        private const val BASE_URL = "https://api.prevengos.test/"
    }
}
